README.md
# Introduction

Pydoxine is an isomer of Pyridoxine, vitamin B6, that allows you to put some
vitamins in your Boost.Python wrappings.

Pydoxine doesn't come in pellets, but is a collection of Python and CMake
scripts that solves the problem of resusing the doxygen documentation of
C++ headers to use it as docstrings in the wrapping. The main idea is to use
the xml output from doxygen to transform it into text strings that can be plugged
in the wrapping definitions. This transformation is carried out by a tweaked
version of [breathe](https://github.com/michaeljones/breathe), an extension for
the documentation system Sphinx.

From a user point of view the key feature provided by this project is a CMake
function with all the logic required to generate the C++ files to be used
to provide docstrings inside your Boost.Python code.

# Dependencies

CMake >= 3
python-docutils
sphinx >= 1.3

The modified version of breathe is shipped as part of these sources.

# Instructions

The following guide provides the instructions to integrate and use Pydoxine
inside an existint project. This guide assumes the reader is familiar with
CMake and git.

## Code base integration

To use Pydoxine, first you have to integrate in the code base of you project.
The instruction to do it using git submodules follow:

 1. From the repository tree:

        $ git submodule add https://github.com/BlueBrain/Pydoxine docstrings

    Here we have choosen docstrings as the target directory, but feel free to
    change it to whatever suits your needs.

 2. And then commit the addition of the submodule. If you index was clean you
    can simply do:
 commit -am "Addition of Pydoxine git module"

## CMake modifications

Assuming you have imported Pydoxine under ${PROJECT_SOURCE_DIR}/docstrings,
the code that you need to add to your CMake scripts is this:

    list(APPEND CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/docstrings/CMake)
    include(PythonDocstrings)

    docstrings(WRAPPING_SOURCES WRAPPED_HEADERS)

where `WRAPPED_SOURCES' is a CMake variable with a list of Boost.Python C++
sources and `WRAPPED_HEADERS' is another one with the header source files for the
doxygen documentation. One limitation of the current implementation is that the
header and source file lists must use absolute paths.

The macro `docstrings' will parse the Boost.Python sources and produce a file
called init_docstrings.cpp in ${PROJECT_BINARY_DIR}/docstrings/cpp. This file
is included from ${PROJECT_BINARY_DIR}/docstrings/cpp/docstrings.cpp.

The CMake target that builds your Python module must include
docstrings/cpp/docstrings.cpp as a source file and it must be made depend
on ${PROJECT_NAME}-docstrings (a target created by the macro docstrings):

    add_dependencies(your_target ${PROJECT_NAME}-docstrings)

## Boost.Python sources modifications

In your Boost.Python sources you have to include the header file
${PROJECT_BINARY_DIR}/docstrings/cpp/docstrings.h.

This header defines the macros DOXY_CLASS, DOXY_ENUM and DOXY_FN. These macros
generate the code that obtains the C-style string to be used as docstring
for the wrapping. The argument taken is always the fully qualified name of the
object to be documented. For overloaded functions you can choose a particular
overload including the function signature. If no signature is provided, the
docstring will contain the documentation of all the overloads (and the output
text won't be rendered correctly inside the Python console).

For optimal results include this line at the beginning of the module:
  boost::python::docstring_options doc_options(true, true, false);

# Example

The directory called example contains a full project showing how all the pieces
fit together.


doc/Changelog.md
Changelog {#Changelog}
=========

# git master

# Release 1.0 (09-12-2016)

* [7](https://github.com/BlueBrain/Pydoxine/pull/7):
  * Added the alias for the custom Doxygen directive \env.
* [3](https://github.com/BlueBrain/Pydoxine/pull/3):
  * Changes in Doxygen config to support better full macro expansion.
  * Added a new input variable to provide include paths to Doxygen.
* Initial version


example/example.h
/* Copyright (c) 2016, EPFL/Blue Brain Project
 *                     Juan Hernando <jhernando@fi.upm.es>
 *
 * This file is part of Pydoxine <https://github.com/BlueBrain/Pydoxine>
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License version 3.0 as published
 * by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifndef EXAMPLE
#define EXAMPLE

#include <string>

namespace example
{

/**
    \brief A class with a short description that is longer than normal to
    test how the final layout looks like.

    A long description: Sed ut perspiciatis unde omnis iste natus error sit
    voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque
    ipsa quae ab illo inventore veritatis et quasi architecto beatae vitae
    dicta sunt explicabo. Nemo enim ipsam voluptatem quia voluptas sit
    aspernatur aut odit aut fugit, sed quia consequuntur magni dolores eos qui
    ratione voluptatem sequi nesciunt. Neque porro quisquam est, qui dolorem
    ipsum quia dolor sit amet, consectetur, adipisci velit, sed quia non
    numquam eius modi tempora incidunt ut labore et dolore magnam aliquam
    quaerat voluptatem.
*/
class BaseExample
{
public:
    /**
       A list of things
    */
    enum Things
    {
    #ifdef DOXYGEN_TO_BREATHE
        FOO = 0, //!< A foo thing
        BAR = 1 //!< A bar thing
    #else
        THING_FOO = 0, //!< A foo thing
        THING_BAR = 1 //!< A bar thing
    #endif
    };

    /** A static class member. */
    static Things theThing;
};

namespace module
{

/** A free enum with just one value */
enum Enum { ENUM_ONE };

/**
\brief This is the brief

@ingroup Main

And this the long
  - hola
  - lista
*/
class Example : public BaseExample
{
public:
    /**
       A nested class.

       This is an opaque type.
    */
    class Nested;

    /** \brief The pulic method */
    void publicMethod() {}

    /**
        An overload of the public method with arguments.

        This one has 2 arguments

        @param a One parameter
        @param b Other parameter

        @see publicMethod
    */
    void publicMethod(int a, std::string b) {}

    /** \brief A method without much to say but a brief comment two lines
     long (in the final output).

     Rather uninteresting.

     @param x
     @return int
    */
    int f(int x) const { return 0; }

protected:
    /** The protected method */
    void protectedMethod() {}
};

}

}
#endif


breathe/doxygen_mangling.py
import re
from six import StringIO

def unmangleId(id) :
    """
    Unmangles a doxygen ID and returns a tuple with the scope type, the
    C++ identifier of the scope (empty for files, which means global scope)
    and the identified type. When the identified type is a unique ID the
    returned id is an empty string (the name will have to be retrieved from
    the 'name' property of the data object with this ID).
    """
    out = StringIO()
    underscore = False
    c = 0
    while c < len(id) :
        if id[c] != '_' :
            if underscore :
                if id[c] == '1' :
                    out.write(':')
                elif id[c] == '8' :
                    out = StringIO()
                    out.write('file')
                    # Skipping extension
                    c += 1
                    if id[c] != 'h' :
                        c += 2
                else :
                    out.write(id[c].upper())
            else :
                out.write(id[c])
            underscore = False
        else :
            if underscore :
                out.write('_')
            underscore = not underscore
        c += 1
    out.seek(0)
    out = out.read()

    for t in ['class', 'struct', 'namespace', 'file'] :
        if t in out[:len(t)] :
            id = out[len(t):]
            if re.search("([^:]|^):[^:]", id) :
                scope = id.rsplit(':', 1)[0]
                return (t, scope, '')
            else :
                try :
                    scope, id = id.rsplit('::', 1)
                    return (t, scope, id)
                except ValueError:
                    return (t, '', id)

    #This is a file, removing the file name

    return out

def mangleId(id) :
    """
    Mangles a C++ identifier as a doxygen doxygen ID without any prefix
    """
    out = StringIO()
    for c in id :
        if c == '_' :
            out.write('__')
        elif c == ':' :
            out.write('_1')
        elif c.isupper() :
            out.write('_')
            out.write(c.lower())
        else :
            out.write(c)
    out.seek(0)
    return out.read()


docstrings/sphinx_ext/cpp_docstrings.py
# Copyright (c) 2016, EPFL/Blue Brain Project
#                     Juan Hernando <jhernando@fi.upm.es>
#
# This file is part of Pydoxine <https://github.com/BlueBrain/Pydoxine>
#
# This library is free software; you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License version 3.0 as published
# by the Free Software Foundation.
#
# This library is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

# Code based in a similar script from the OpenRAVE project.

from sphinx.builders import Builder
import sphinx.addnodes
from six import StringIO
import xml2rst
import re
import os
import sys

string_type = str if  sys.version_info.major == 3 else unicode

xml2rst.setDefaultOptions()
mainXsltF = open(os.path.dirname(__file__) + '/xml2rst.xsl','r')

def mangleCPPIdentifier(id) :
    out = StringIO.StringIO()
    for t in ['class', 'function', 'enum'] :
        if t in id[:len(t)] :
            id = id[len(t):]
    underscore = False
    for c in id :
        if c != '_' :
            if underscore :
                if c == '1' :
                    out.write(':')
                else :
                    out.write(c.upper())
            else :
                out.write(c)
            underscore = False
        else :
            if underscore :
                out.write('_')
            underscore = not underscore
    return out.read()

def append_text(output, nodes):
    if type(nodes) == sphinx.addnodes.desc:
        assert(type(nodes[1]) == sphinx.addnodes.desc_content)
        nodes = nodes[1]

    whitespace = re.compile('^[\s]*$')
    breathe_directive = re.compile('[\s]*\.\. class:: [a-z]+[\s]*$')

    for node in nodes:
        # Documentation of enum values is really special because they sphinx
        # rendering intersperses this desc and index nodes that xml2rst can't
        # handle.
        # The following code tries to deal with that.
        if type(node) == sphinx.addnodes.index:
            # This can be ignored, we don't need it at all
            continue
        if type(node) == sphinx.addnodes.desc:
            # This is the enum value name
            desc_name = node[0][1]
            assert(type(desc_name) == sphinx.addnodes.desc_name)
            output.write(" * " + desc_name[0])
            # And this is the description
            if len(node[1]) != 0:
                output.write(": ")
            desc_content = node[1]
            assert(type(desc_content) == sphinx.addnodes.desc_content)
            append_text(output, desc_content)
            continue

        domtree=node.asdom()
        inF = StringIO()
        domtree.writexml(inF, indent="", addindent="", newl="")
        inF.seek(0)

        outF = StringIO()
        xml2rst.convert(inF=inF, outF=outF, mainXsltF=mainXsltF)
        outF.seek(0)
        empty_line = False

        comment = StringIO()
        empty_comment = True
        for line in outF.read().split('\n') :
            if breathe_directive.match(line):
                continue

            if empty_line and not empty_comment :
                comment.write('\\n')
            empty_line = False
            empty_comment = False
            # Quoting some characters
            line = re.sub('\\\\','\\\\\\\\', line)
            line = re.sub('\t','\\\\t', line)
            line = re.sub('"','\\\\"', line)
            line = re.sub('([a-zA-Z])::([a-zA-Z])','\\1.\\2', line)

            comment.write(line)
            comment.write('\\n')
        comment.seek(0)
        output.write(comment.read())
        inF.close()
        outF.close()



class CDocstringBuilder(Builder) :
    """
    Builds C style strings with the docstrings to use in Boosth Python
    """
    name = 'docstrings'
    format = 'docstrings'

    def __init__(self, app) :
        super(CDocstringBuilder, self).__init__(app)

    def get_relative_uri(self, docname, typ=None):
        # This is irrelevant
        return docname

    def get_outdated_docs(self) :
        return 'all'

    def init(self) :
        pass

    def prepare_writing(self, docnames) :
        pass

    def write_doc(self, docname, doctree) :

        comments = StringIO()
        # Sorting the names alphabetically to minimize changes over reviews.
        nodes = []
        for node in doctree:
            if not 'fqname' in node :
                # Skipping node without name
                continue
            nodes.append((node['fqname'], node))
        nodes.sort()

        for name, node_stack in nodes:
            comments.write('m["%s"] = "' % name)
            append_text(comments, node_stack[0])
            comments.write('";\n')

        out = open(self.app.config.cpp_docstrings_out_file, 'w')
        out.write("""\
// File automatically generated by %s

void initDocStrings(std::map<std::string, std::string> &m)
{
""" % __file__)
        comments.seek(0)
        out.write(comments.read())
        out.write("}\n")

    def finish(self) :
        print('finish')

def setup(app) :
    app.add_builder(CDocstringBuilder)
    app.add_config_value('cpp_docstrings_out_file', False, False)


example/python/example.cpp
/* Copyright (c) 2016, EPFL/Blue Brain Project
 *                     Juan Hernando <jhernando@fi.upm.es>
 *
 * This file is part of Pydoxine <https://github.com/BlueBrain/Pydoxine>
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License version 3.0 as published
 * by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <boost/python.hpp>

#include "../example.h"

using namespace boost::python;
using namespace example;

BaseExample::Things BaseExample::theThing;

#include "docstrings.h"

void (module::Example::*Example_publicMethod1)() =
    &module::Example::publicMethod;
void (module::Example::*Example_publicMethod2)(int, std::string) =
    &module::Example::publicMethod;

class dummy_namespace {};

BOOST_PYTHON_MODULE(_example)
{

docstring_options doc_options(true, true, false);

class_<BaseExample> baseExampleWrapper(
    "BaseExample", DOXY_CLASS(example::BaseExample), no_init);
{
    scope baseExampleScope = baseExampleWrapper;

    enum_<BaseExample::Things> things(
        "Things", DOXY_ENUM(example::BaseExample::Things));
    things.value("FOO", BaseExample::THING_FOO);
    things.value("BAR", BaseExample::THING_BAR);

    baseExampleWrapper
        .def_readonly("finished", &BaseExample::theThing,
                      DOXY_VAR(example::BaseExample::theThing));
}

enum_<example::module::Enum>("Enum", DOXY_ENUM(example::module::Enum))
    .value("ONE", example::module::ENUM_ONE);

scope module = class_<dummy_namespace>("module", no_init);

class_<example::module::Example> exampleWrapper(
    "Example", DOXY_CLASS(example::module::Example), init<>());
{
    scope exampleScope = exampleWrapper;

    exampleWrapper
        .def("publicMethod", Example_publicMethod1,
             DOXY_FN(example::module::Example::publicMethod()))
        .def("publicMethod", Example_publicMethod2,
             DOXY_FN(example::module::Example::publicMethod(int, std::string)))
        .def("f", &module::Example::f,
             DOXY_FN(example::module::Example::f))
        ;
}

}
